### PR TITLE
#589: Include a blocks tab within content admin page options

### PR DESCRIPTION
**Pull request by barrettbaffert**
### Motivation
Currently, you can only view and edit blocks by navigating to Structure > Blocks > Custom block library. It would be nice to have the same experience as QS1 by way of Content > Blocks navigation.

### Solution
Add a new display for Blocks to the admin view displaying the Content page.

After below:

image

### Concern
I haven't yet found a non-hacky way to add an `Add Custom Block` button above the view of blocks when path is `/admin/content/blocks`. Current solution is hard coded html as a header for the view. 🤮 

## Related Issue

Fixes https://github.com/az-digital/az_quickstart/issues/589
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I've added this Pull Request to the AZ Quickstart Github project and set it to "Review in progress".
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by smayden** [comment]
Initial testing looks good for azadmin, but can this be added for other roles? (definitely azcontentadmin, probably azcontenteditor)

**Comment by danahertzberg** [comment]
> Initial testing looks good for azadmin, but can this be added for other roles? (definitely azcontentadmin, probably azcontenteditor)

Agreed, this should be allowed for content editors

**Comment by danahertzberg** [comment]
Once Shoshana's permissions changes are merged into this PR, you can update the view to use the permission of Edit Flexible Blocks instead of Administer Blocks

**Comment by barrettbaffert** [comment]
Permissions for azcontenteditor should be working as expected.

Next step:

[ ] Figure out how to make local action block with `Add Custom Block` button appear at `/admin/content/blocks`

For reference, here's the [local actions block in core](https://git.drupalcode.org/project/drupal/-/blob/9.2.x/core/lib/Drupal/Core/Menu/Plugin/Block/LocalActionsBlock.php).

**Comment by barrettbaffert** [comment]
> * [x]   Figure out how to make local action block with `Add Custom Block` button appear at `/admin/content/blocks`

Because @trackleft is too good to be and can create magic, this task is now complete and I believe this Pull Request is ready for full review

**Review by smayden — COMMENTED** [review]
Tested as content editor and content admin and working as expected.

One suggestion I think we should consider before merging is to add a "Reset" button to the filter on the view. I think you just can change line 40 to `reset_button: true` for the views.view.block_content.yml file.

**Review by trackleft — DISMISSED** [review]

File: config/install/views.view.block_content.yml [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+          disable_sql_rewrite: false
+          distinct: false
+          replica: false
+          query_comment: ''
+          query_tags: {  }
+      exposed_form:
+        type: basic
+        options:
+          submit_button: Apply
+          reset_button: false
```
Comment (barrettbaffert) [inline config/install/views.view.block_content.yml:40]:
```suggestion
          reset_button: true
```

File: modules/custom/az_flexible_block/src/Plugin/Menu/LocalAction/AzFlexibleBlockAddLocalAction.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getOptions(RouteMatchInterface $route_match) {
+    $options = parent::getOptions($route_match);
+    // If the route specifies a theme, append it to the query string.
+    if ($theme = $route_match->getParameter('theme')) {
+      $options['query']['theme'] = $theme;
+    }
```
Comment (trackleft) [inline modules/custom/az_flexible_block/src/Plugin/Menu/LocalAction/AzFlexibleBlockAddLocalAction.php:22]: As long as we are making updates...  Probably don't need this part

```suggestion
```

File: config/install/views.view.block_content.yml [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+    cache_metadata:
+      contexts:
+        - 'languages:language_content'
+        - 'languages:language_interface'
+        - url
+        - url.query_args
+        - user.permissions
+      max-age: -1
+      tags: {  }
+  page_1:
```
Comment (trackleft) [inline config/install/views.view.block_content.yml:516]:
```suggestion
  admin:
```

File: config/install/views.view.block_content.yml [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+    display_plugin: page
+    id: page_1
```
Comment (trackleft) [inline config/install/views.view.block_content.yml:518]:
```suggestion
    id: admin
```

File: modules/custom/az_flexible_block/az_flexible_block.links.action.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+az_flexible_block_add_action:
+  route_name: block_content.add_page
+  title: 'Add custom block'
+  appears_on:
+    - view.block_content.page_1
```
Comment (trackleft) [inline modules/custom/az_flexible_block/az_flexible_block.links.action.yml:5]:
```suggestion
    - view.block_content.admin
```

**Review by trackleft — DISMISSED** [review]
And I would prefer a better machine name, personally.

**Comment by smayden** [comment]
I'm seeing a new issue when I test as content admin and content editor that is different than when I tested previously. When I now go to add a custom block using the button, once I hit save it now seems to go to the block configuration (previously I think it just went back to the list of blocks).

For content editors this gives an access designed message. Not sure what might have changed here though?

image

The block is successfully created and everything else working as expected

File: modules/custom/az_flexible_block/src/Plugin/Menu/LocalAction/AzFlexibleBlockAddLocalAction.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ */
+class AzFlexibleBlockAddLocalAction extends LocalActionDefault {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getOptions(RouteMatchInterface $route_match) {
+    $options = parent::getOptions($route_match);
+    // Adds a destination on custom block listing.
+    if ($route_match->getRouteName() == 'view.block_content.page_1') {
```
Comment (barrettbaffert) [inline modules/custom/az_flexible_block/src/Plugin/Menu/LocalAction/AzFlexibleBlockAddLocalAction.php:20]:
```suggestion
    if ($route_match->getRouteName() == 'view.block_content.admin') {
```
@trackleft do I need to make this change too?

**Review by smayden — APPROVED** [review]
Tested as content editor and content admin. All working beautifully now!

**Review by danahertzberg — APPROVED** [review]

**Review by mmunro-ltrr — APPROVED** [review]

**Review by trackleft — APPROVED** [review]

**Pull request merged**